filesystem: introduce swap

romain-dartigues opened the pull request:
##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
- filesystem

##### ANSIBLE VERSION
```
ansible 2.1.2.0
```

##### SUMMARY
- 4266b1a aggregate failures (minor code design, shared with PR #3313)
- c7ae654 introduce swap (new feature!)

I consider the swap as a new feature _and_ a bug.

Before the patch (`/dev/DATA/swap` is an empty, newly created LV):

``` yaml
- name: mkswap
  filesystem: fstype=swap dev=/dev/DATA/swap
```

Returned:

```
ok: [docker02] => (item={u'fs_vfstype': u'swap', u'fs_spec': u'/dev/DATA/swap', u'size': u'2G'})
```

But on the target, nothing much happened:

``` console
root@docker02:~# blkid /dev/DATA/swap
root@docker02:~# file -Ls /dev/DATA/swap
/dev/DATA/swap: data
```

After the patch, the playbook says:

```
changed: [docker02] => (item={u'fs_vfstype': u'swap', u'fs_spec': u'/dev/DATA/swap', u'size': u'2G'})
```

Now on the target:

``` console
root@docker02:~# blkid /dev/DATA/swap
/dev/DATA/swap: UUID="40fc8cb4-f18a-4693-85e4-189b52ff66b3" TYPE="swap"
root@docker02:~# file -Ls /dev/DATA/swap
/dev/DATA/swap: Linux/i386 swap file (new style), version 1 (4K pages), size 524287 pages, no label, UUID=40fc8cb4-f18a-4693-85e4-189b52ff66b3
```

Thanks to the existing code, the swap functionality is idempotent and when re-run will detect the FS is already a swap:

```
ok: [docker02] => (item={u'fs_vfstype': u'swap', u'fs_spec': u'/dev/DATA/swap', u'size': u'2G'})
```

The target is:

```
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 14.04.4 LTS
Release:        14.04
Codename:       trusty
```

**Note:** split from #3313.

mscherer commented on system/filesystem.py:
```diff
@@ -101,6 +106,20 @@ def _get_fs_size(fssize_cmd, dev, module):
+            for line in size.splitlines():
+                row = line.split()
+                if row[0] == dev:
+                    return int(row[2]) * 1024
```
I think it would be better to not add a 'return' here when the other part part of the code don't.

And if the swap device is not in us, it will fail, and continue to 'return block_size*block_count' with 2 unassigned variable.

romain-dartigues commented:
> And if the swap device is not in us, it will fail, and continue to 'return block_size*block_count' with 2 unassigned variable.

That's my fault when I split #3313, where I changed `block_size` and `block_count` to be always defined.

If you prefer I will not return line 120 (I personally prefer to exit early).

mscherer commented:
Both work, but I think being coherent is better. But at the same time, I am also not the maintainer of the module, and I do not want to be annoying. The rest of the code did look fine.

gregdek commented:
Thanks @romain-dartigues. To the current maintainers, @abulimov please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

mscherer commented:
For the record, i would be interested into this feature

duck-rh commented:
any news? without this this module is really incomplete.

gregdek commented:
Thanks @romain-dartigues for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

For help on how to do this cleanly please see http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

[This message brought to you by your friendly Ansibull-bot.]

romain-dartigues commented:
ready_for_review

gregdek commented:
Thanks @romain-dartigues. To the current maintainers, @abulimov please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]
